src/neo/SmartContract/ApplicationEngine.cs

```diff
@@ -67,6 +69,7 @@ protected ApplicationEngine(TriggerType trigger, IVerifiable container, StoreVie
             this.ScriptContainer = container;
             this.Snapshot = snapshot;
             this.gas_amount = gas;
+            this.base_exec_fee = snapshot is null ? PolicyContract.DefaultBaseExecFee : NativeContract.Policy.GetBaseExecFee(Snapshot);
```

Comment:
I think the `snapshot` is impossible to be `null`.

Reply:
In some cases i.e Transaction.VerifyIndependent snapshot can be null.

Reply:
I think normally yes, it shouldn't be `null`, but at least some unit tests fail without this check.

Reply:
There are two cases where this is `null`: unit tests (those are fine with this code) and `Transaction.VerifyStateIndependent`. `VerifyStateIndependent` is a problem (it needs to calculate GAS properly) and I have fixed that, although it makes things a bit more complicated (the other option probably is to use `Blockchain.Singleton.GetSnapshot()`).

Reply:
> In some cases i.e Transaction.VerifyIndependent snapshot can be null.

Then it is impossible to have `VerifyIndependent`. Because the `base_exec_fee` is a state to be read from the `snapshot`.

Reply:
> > In some cases i.e Transaction.VerifyIndependent snapshot can be null.
> 
> Then it is impossible to have `VerifyIndependent`. Because the `base_exec_fee` is a state to be read from the `snapshot`.

Func `VerifyIndependent` is introduced to verify transaction witness parallelly to enhance TPS, without such strategy TPS would drop severely.

Reply:
> 
> Func `VerifyIndependent` is introduced to verify transaction witness parallelly to enhance TPS, without such strategy TPS would drop severely.

We made TPS benchmark for the [adjustable_opcode_prices](https://github.com/roman-khimov/neo/tree/adjustable-opcode-prices) against [1d4db2c from master branch](https://github.com/neo-project/neo/commit/1d4db2cd4176390d18deeffc2aacc230a55950b0) under the load of 30 working threads pushing NEP-5 transfer transactions to the single-node network. It almost do not affect average TPS value, see the results below:
| Revision | Average TPS |
| --- | --- |
| [master](https://github.com/neo-project/neo/commit/1d4db2cd4176390d18deeffc2aacc230a55950b0) | 3285.665 | 
| [adjustable_opcode_prices](https://github.com/roman-khimov/neo/tree/adjustable-opcode-prices) | 3305.576 |

![tps_single_30_wrk](https://user-images.githubusercontent.com/22775709/98528767-d0939c80-228d-11eb-8ce5-ef8c532ee484.png)
![tpb_single_30_wrk](https://user-images.githubusercontent.com/22775709/98528782-d4bfba00-228d-11eb-9324-c8cb940c3514.png)
![ms_per_block_single_30_wrk](https://user-images.githubusercontent.com/22775709/98528793-d8ebd780-228d-11eb-8159-62bfb1bec29a.png)
![mem_single_30_wrk](https://user-images.githubusercontent.com/22775709/98528798-dbe6c800-228d-11eb-907d-c406dbf31836.png)
![cpu_single_30_wrk](https://user-images.githubusercontent.com/22775709/98528806-de492200-228d-11eb-944b-32fc95c176a5.png)

Reply:
@AnnaShaleva 

(1) This assertion is against siituation where func `VerifyIndependent ` is dropped, or,

(2) Before https://github.com/neo-project/neo/pull/2045/commits/2038bd0e895f19a34b987a24a14aa64050acfd42 is applied

So your TPS result might be good, but it's not what I mean.

Moreover, @roman-khimov passing `currentSnapshot` to parellel reading is not safe & correct, especially it will be refreshed during func `Persist`

Reply:
It's after 2038bd0 and it at least shows that concurrent read-only snapshot access is not a problem performance-wise (and these changes are not a problem protocol-wise). But I agree that for current C# implementation it's important for `VerifyIndependent` to not really have any dependence on snapshots. What really technically is needed is for all verifications to be done using the same context (even `VerifyStateDependent` can be done concurrently, because the state is inherently RO in verification context), this is the way it's done in neo-go since 0.72.0. Now to how to properly implement this model here is a bit harder question.

Reply:
>  but at least some unit tests fail without this check.

Then I think that we should fix the UT

Reply:
Good point, but I think we need to solve `snapshot` problem first, I'll fix them afterwards.

Reply:
BTW, we have now cases where snapshot is explicitly `null` (state-independent verification), so I think we still need this null check.